redwood-treehouse/src/hostMain/kotlin/app/cash/redwood/treehouse/EventListener.kt

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package app.cash.redwood.treehouse
+
+import app.cash.zipline.Call
+import app.cash.zipline.CallResult
+import app.cash.zipline.ZiplineService
+
+public abstract class EventListener {
+  public open fun appCreated(
+    app: TreehouseApp<*>,
+  ) {
+  }
+
+  public open fun appCanceled(
+    app: TreehouseApp<*>,
+  ) {
+  }
+
+  public open fun codeLoadStart(
+    app: TreehouseApp<*>,
+    manifestUrl: String?,
+  ): Any? = Unit
+
+  public open fun codeLoadEnd(
+    app: TreehouseApp<*>,
+    manifestUrl: String?,
+    startValue: Any?,
+  ) {
+  }
+
+  public open fun codeLoadFailed(
+    app: TreehouseApp<*>,
+    manifestUrl: String?,
+    exception: Exception,
+    startValue: Any?,
+  ) {
+  }
+
+  public open fun onUnknownWidget(
+    app: TreehouseApp<*>,
+    kind: Int,
+  ) {
+  }
+
+  public open fun onUnknownLayoutModifier(
+    app: TreehouseApp<*>,
+    tag: Int,
+  ) {
+  }
+
+  public open fun onUnknownChildren(
+    app: TreehouseApp<*>,
+    kind: Int,
+    tag: Int,
+  ) {
+  }
+
+  public open fun onUnknownProperty(
+    app: TreehouseApp<*>,
+    kind: Int,
+    tag: Int,
+  ) {
+  }
+
+  public open fun downloadStart(
+    app: TreehouseApp<*>,
+    url: String,
+  ): Any? = Unit
+
+  public open fun downloadEnd(
+    app: TreehouseApp<*>,
+    url: String,
+    startValue: Any?,
+  ) {
+  }
+
+  public open fun downloadFailed(
+    app: TreehouseApp<*>,
+    url: String,
+    exception: Exception,
+    startValue: Any?,
+  ) {
+  }
+
+  public open fun manifestParseFailed(
+    app: TreehouseApp<*>,
+    url: String?,
+    exception: Exception,
+  ) {
+  }
+
+  public open fun bindService(
+    name: String,
+    service: ZiplineService,
+  ) {
+  }
+
+  public open fun takeService(
+    name: String,
+    service: ZiplineService,
+  ) {
+  }
+
+  public open fun callStart(
+    call: Call,
+  ): Any? = Unit
+
+  public open fun callEnd(
```

Comment:
should we have a `callFailed`? or is that going to be a part of the CallResult? Either way, will we have an array of the stacktrace here?

Reply:
Yep - `CallResult` provides a `kotlin.Result` so you can do `result.result.exceptionOrNull()` to get the exception if the call failed.